modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp

```diff
@@ -0,0 +1,306 @@
+            {
+                do {
+                    //TODO try to extract TIMESTAMP
+                    sess.last_status = MFXVideoCORE_SyncOperation(sess.session, sess.sync, 100);
+                    if (MFX_ERR_NONE == sess.last_status) {
+
+                        LegacyDecodeSession& my_sess = static_cast<LegacyDecodeSession&>(sess);
+                        on_frame_ready(my_sess);
+                    }
+                } while (sess.last_status == MFX_WRN_IN_EXECUTION);
+            }
+            return ExecutionStatus::Continue;
+        },
+        // 4) Falls back on generic status procesing
+        [this] (EngineSession& sess) -> ExecutionStatus
+        {
+            return this->process_error(sess.last_status, static_cast<LegacyDecodeSession&>(sess));
+        }
+    );
+}
+
+void VPLLegacyDecodeEngine::initialize_session(mfxSession mfx_session,
+                                         DecoderParams&& decoder_param,
+                                         std::shared_ptr<IDataProvider> provider)
+{
+    mfxFrameAllocRequest decRequest = {};
+    // Query number required surfaces for decoder
+    MFXVideoDECODE_QueryIOSurf(mfx_session, &decoder_param.param, &decRequest);
+
+    // External (application) allocation of decode surfaces
+    GAPI_LOG_DEBUG(nullptr, "Query IOSurf for session: " << mfx_session <<
+                            ", mfxFrameAllocRequest.NumFrameSuggested: " << decRequest.NumFrameSuggested <<
+                            ", mfxFrameAllocRequest.Type: " << decRequest.Type);
+
+    mfxU32 singleSurfaceSize = GetSurfaceSize(decoder_param.param.mfx.FrameInfo.FourCC,
+                                              decoder_param.param.mfx.FrameInfo.Width,
+                                              decoder_param.param.mfx.FrameInfo.Height);
+    if (!singleSurfaceSize) {
+        throw std::runtime_error("Cannot determine surface size for: fourCC" +
+                                 std::to_string(decoder_param.param.mfx.FrameInfo.FourCC) +
+                                 ", width: " + std::to_string(decoder_param.param.mfx.FrameInfo.Width) +
+                                 ", height: " + std::to_string(decoder_param.param.mfx.FrameInfo.Height));
+    }
+
+    const auto &frameInfo = decoder_param.param.mfx.FrameInfo;
+    auto surface_creator =
+            [&frameInfo] (std::shared_ptr<void> out_buf_ptr, size_t out_buf_ptr_offset,
+                          size_t out_buf_size) -> surface_ptr_t {
+                return (frameInfo.FourCC == MFX_FOURCC_RGB4) ?
+                        create_surface_RGB4(frameInfo, out_buf_ptr, out_buf_ptr_offset,
+                                            out_buf_size) :
+                        create_surface_other(frameInfo, out_buf_ptr, out_buf_ptr_offset,
+                                             out_buf_size);};
+
+    //TODO Configure preallocation size (how many frames we can hold)
+    const size_t preallocated_frames_count = 30;
```

Comment:
Have we talked about this magic number before or it was another one? If so, is there any way to make it configurable from the outside?